PyDSS/pyPostprocessor/PostprocessScripts/AutomatedThermalUpgrade.py

```diff
@@ -431,6 +496,188 @@ def read_available_upgrades(self, file):
+        self.buses_with_violations = []
+        self.buses_with_undervoltage_violations = []
+        self.buses_with_overvoltage_violations = []
+        self.buses_with_violations_pos = {}
+        self.nodal_violations_dict = {}
+        # If multiple load files are being used, the 'tps_to_test property is not used, else if a single load file is
+        # used use the 'tps to test' input
+        if len(self.other_load_dss_files)>0:
+            for tp_cnt in range(len(self.config["tps_to_test"])):
+                # First two tps are for disabled PV case
+                if tp_cnt == 0 or tp_cnt == 1:
+                    dss.run_command("BatchEdit PVSystem..* Enabled=False")
+                    dss.Loads.First()
+                    while True:
+                        load_name = dss.Loads.Name().split(".")[0].lower()
+                        if load_name not in self.min_max_load_kw:
+                            raise Exception("Load not found, quitting...")
+                        new_kw = self.min_max_load_kw[load_name][tp_cnt]
+                        dss.Loads.kW(new_kw)
+                        if not dss.Loads.Next()>0:
+                            break
+                    self.dssSolver.Solve()
+                    if not dss.Solution.Converged():
+                        self.logger.info("Write upgrades before Convergence Error in debug_upgrades.dss")
+                        self.write_upgrades_to_file(
+                            output_path=os.path.join(self.config["Outputs"], "debug_upgrades.dss"))
+                        if raise_exception:
+                            raise OpenDssConvergenceError("OpenDSS solution did not converge")
+                            return False
+                        else:
+                            return False
+                if tp_cnt == 2 or tp_cnt == 3:
+                    dss.run_command("BatchEdit PVSystem..* Enabled=True")
+                    dss.Loads.First()
+                    while True:
+                        load_name = dss.Loads.Name().split(".")[0].lower()
+                        if load_name not in self.min_max_load_kw:
+                            raise Exception("Load not found, quitting...")
+                        new_kw = self.min_max_load_kw[load_name][tp_cnt-2]
+                        dss.Loads.kW(new_kw)
+                        if not dss.Loads.Next() > 0:
+                            break
+                    self.dssSolver.Solve()
+                    if not dss.Solution.Converged():
+                        self.logger.info("Write upgrades before Convergence Error in debug_upgrades.dss")
+                        self.write_upgrades_to_file(
+                            output_path=os.path.join(self.config["Outputs"], "debug_upgrades.dss"))
+                        if raise_exception:
+                            raise OpenDssConvergenceError("OpenDSS solution did not converge")
+                            return False

```

Comment:
here, also. And several other places below.

Reply:
(y)